src/data/localization.rs

```diff
@@ -537,29 +503,54 @@ impl Localization {
         }
     }
 
+    fn validate_loca(
+        entry: &LocaEntry,
+        from: &TigerHashMap<String, LocaEntry>,
+        data: &Everything,
+        sc: &mut ScopeContext,
+        lang: &'static str,
+    ) {
+        if matches!(entry.value, LocaValue::Macro(_)) {
+            let mut new_line = Vec::new();
+            let mut count = 0;
+            if entry.expand_macros(&mut new_line, from, &mut count, sc, None) {
+                // re-parse after macro expansion
+                let new_line_as_ref = new_line.iter().collect();
+                let mut value = ValueParser::new(new_line_as_ref).parse_value();
+                let value = if value.len() == 1 {
+                    std::mem::take(&mut value[0])
+                } else {
+                    LocaValue::Concat(value)
+                };
+                Self::check_loca_code(&value, data, sc, lang);
+            }
+        } else {
+            Self::check_loca_code(&entry.value, data, sc, lang);
+        }
+    }
+
     pub fn validate_use(&self, key: &str, data: &Everything, sc: &mut ScopeContext) {
-        self.keys_validated_with_sc.write().unwrap().insert(key.to_string());
         for lang in &self.mod_langs {
             if let Some(hash) = self.locas.get(lang) {
                 if let Some(entry) = hash.get(key) {
-                    Self::check_loca_code(&entry.value, data, sc, lang);
+                    entry.used.store(true, Relaxed);
+                    entry.validated.store(true, Relaxed);
+                    Self::validate_loca(entry, hash, data, sc, lang);
                 }
             }
         }
     }
 
-    // This is in pass2 to make sure all `keys_validated_with_sc` have been marked.
+    // This is in pass2 to make sure all `validated` entries have been marked.
     pub fn validate_pass2(&self, data: &Everything) {
         scope(|s| {
-            // Hold the lock for the whole validation loop, to avoid the overhead of re-acquiring it
-            let already_validated = self.keys_validated_with_sc.read().unwrap();
             for (lang, hash) in &self.locas {
                 for entry in hash.values() {
-                    if !already_validated.contains(entry.key.as_str()) {
+                    if !entry.validated.load(Relaxed) {
                         s.spawn(|_| {
                             let mut sc = ScopeContext::new_unrooted(Scopes::all(), &entry.key);
```

Comment:
maybe a good idea to comment out `entry.validated.store(true, Relaxed)` but note it's not needed after this anyway?

Reply:
Yeah I added a comment